certora/KilnBase.spec

```diff
@@ -0,0 +1,316 @@
+rule deny_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    deny@withrevert(e, usr);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify correct storage changes for not reverting file
+rule file(bytes32 what, uint256 data) {
+    env e;
+
+    uint256 lotBefore = lot();
+    uint256 hopBefore = hop();
+
+    file(e, what, data);
+
+    uint256 lotAfter = lot();
+    uint256 hopAfter = hop();
+
+    assert(what == 0x6c6f740000000000000000000000000000000000000000000000000000000000 => lotAfter == data,      "assert1 failed");
+    assert(what != 0x6c6f740000000000000000000000000000000000000000000000000000000000 => lotAfter == lotBefore, "assert2 failed");
+    assert(what == 0x686f700000000000000000000000000000000000000000000000000000000000 => hopAfter == data,      "assert3 failed");
+    assert(what != 0x686f700000000000000000000000000000000000000000000000000000000000 => hopAfter == hopBefore, "assert4 failed");
+}
+
+// Verify revert rules on file
+rule file_revert(bytes32 what, uint256 data) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    file@withrevert(e, what, data);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+    bool revert3 = what != 0x6c6f740000000000000000000000000000000000000000000000000000000000 && // what is not "lot"
+                   what != 0x686f700000000000000000000000000000000000000000000000000000000000;   // what is not "hop"
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+    assert(revert3 => lastReverted, "revert3 failed");
+
+    assert(lastReverted => revert1 || revert2 || revert3, "Revert rules are not covering all the cases");
+}
+
+// Verify correct storage changes for not reverting rug
+rule rug(address dst) {
+    env e;
+
+    require(dai == sell());
+
+    uint256 balanceKilnBefore = dai.balanceOf(currentContract);
+    uint256 balanceDstBefore = dai.balanceOf(dst);
+    uint256 supplyBefore = dai.totalSupply();
+    bool dstSameAsKiln = currentContract == dst;
+
+    rug(e, dst);
+
+    uint256 balanceKilnAfter = dai.balanceOf(currentContract);
+    uint256 balanceDstAfter = dai.balanceOf(dst);
+    uint256 supplyAfter = dai.totalSupply();
+
+    assert(supplyAfter == supplyBefore, "assert1 failed");
+    assert(!dstSameAsKiln =>
+            balanceKilnAfter == 0 &&
+            balanceKilnBefore == (balanceDstAfter - balanceDstBefore), "assert2 failed");
+    assert(dstSameAsKiln => balanceKilnAfter == balanceKilnBefore, "assert3 failed");
+}
+
+// Verify revert rules on rug
+rule rug_revert(address dst) {
+    env e;
+
+    require(dai == sell());
+
+    uint256 ward = wards(e.msg.sender);
+    uint256 locked = lockedGhost();
```

Comment:
Since there's a public getter for `locked`, I think strictly speaking you don't need to use a ghost, could just do:
```
uint256 locked = locked();
```
(should work for the `fire` revert spec below as well).

Reply:
Yep, tested and confirmed--can just remove the ghost.

Reply:
oh! good catch, yeah I haven't noticed it was public, I thought we were following the old pattern declaring it internal and grabbed the ghost from previous specs, but this is a nice evolution of it and definitely FV-friendly.

Reply:
done in 3a6fb04